### PR TITLE
feat: extract calendar component and refactor scheduling flow

### PR DESCRIPTION
**Pull request by carissa-tang**
## Implementation description. How did you make this change?
* Extracted calendar from input field
* Made form progressive (reveals one section at a time)
* Moved back/next buttons & disable feature added to "next"
* "End date" input field was changed to date picker
* Overall styling was tweaked to match Figma designs
* Note: Was unable to remove "," from Calendar and to only display two weeks -> limitation of the library used (sorry Stacy!)

## Steps to test
1. Click through the form to see if each question is revealed one after the other
2. Ensure all buttons are clickable ("end date" field should only appear if not one time donation)
4.  Cross-check both Desktop and Mobile view with Figma designs

Desktop View: https://user-images.githubusercontent.com/69697180/150065707-0ebb5bdc-8727-47d1-a2ab-bbcc6f5ffffa.mp4
* @hanlinc27 Inconsistency with radio buttons is apparent from 0:24-0:30

Mobile View: https://user-images.githubusercontent.com/69697180/150065486-d421a912-963d-42bb-a499-b04f58014fbb.mov

Update: Steps 2 and 3 are now refactored as well

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages follow conventional commits and are descriptive. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s) 
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR
- [x] The appropriate tests if necessary have been written

### CODE REVIEW

File: backend/typescript/models/scheduling.model.ts [hunk 1]
```diff
@@ -67,7 +67,7 @@ export default class Scheduling extends Model {
 
   @AllowNull(false)
   @Column({
-    type: DataType.ENUM("One time", "Daily", "Weekly", "Monthly"),
+    type: DataType.ENUM("One time donation", "Daily", "Weekly", "Monthly"),
```
Comment (hanlinc27) [inline backend/typescript/models/scheduling.model.ts:70]: can we revert this change (+ the other ones)? there's more places where it's just "One Time" in the platform so you can just map it in the first scheduling page instead of changing it entirely
```suggestion
    type: DataType.ENUM("One time", "Daily", "Weekly", "Monthly"),
```

File: frontend/src/components/common/RadioImageSelectGroup.tsx [hunk 2]
```diff
@@ -38,7 +39,7 @@ const RadioImageSelectButton = (props: any) => {
       cursor: "pointer",
       borderWidth: "1px",
       borderRadius: "6",
-      borderColor: "hubbard.100",
+      borderColor: "#D8DDE0",
```
Comment (hanlinc27) [inline frontend/src/components/common/RadioImageSelectGroup.tsx:42]: can you add this colour as a variant in the `theme.tsx`? Seeing it's used three times

File: frontend/src/components/pages/Scheduling/BackButton.tsx [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+          <Button
+            onClick={onSaveClick}
+            variant="navigation"
+          >
+          Save Changes
+        </Button>
+        ) : (
+        <Button
+          onClick={previous}
+          paddingLeft="0 !important"
```
Comment (hanlinc27) [inline frontend/src/components/pages/Scheduling/BackButton.tsx:19]: let's try to avoid `!important`, since this css property is passed in line, it'll already override and be the highest priority. Usually we use `!important` like you did for custom library stylings to override but should [definitely avoid ](https://uxengineer.com/css-specificity-avoid-important-css/)otherwise! 
```suggestion
          paddingLeft="0"
```

File: frontend/src/components/pages/Scheduling/DonationInformation.tsx [hunk 4]
```diff
@@ -36,6 +38,9 @@ const DonationInformation: any = ({
     size: "",
   });
 
+  const [showItemTypes, setShowItemTypes] = useState<boolean>(false);
```
Comment (hanlinc27) [inline frontend/src/components/pages/Scheduling/DonationInformation.tsx:41]: nit: to be verbose
```suggestion
  const [showDonationItemTypes, setShowDonationItemTypes] = useState<boolean>(false);
```

File: frontend/src/components/pages/Scheduling/DonationInformation.tsx [hunk 5]
```diff
@@ -45,8 +50,20 @@ const DonationInformation: any = ({
...
 
+  const checkSubmit = (e: React.ChangeEvent<HTMLInputElement>, item: string) => {
+    const newCategories = e.target.checked ? [...categories, item] : categories.filter((category) => category !== item);
+    
+    if (newCategories.length > 0) {
+      setCanSubmit(true);
+    } else {
+      setCanSubmit(false);
+    }
+  }
```
Comment (hanlinc27) [inline frontend/src/components/pages/Scheduling/DonationInformation.tsx:65]:
```suggestion
setCanSubmit(!!newCategories.length)
```

File: frontend/src/components/pages/Scheduling/GetStarted.tsx [hunk 6]
```diff
@@ -21,24 +21,26 @@ const GetStarted = ({ navigation }: SchedulingStepProps) => {
         Our goal is to maintain a consistent stock of donated food in the fridge
         & pantry, at any given time, on any given day. The scheduling tool
         enables us to space out donations - effectively maximizing your impact
-        on community
+        on the community.
```
Comment (hanlinc27) [inline frontend/src/components/pages/Scheduling/GetStarted.tsx:24]: nice catch!

File: frontend/src/components/pages/Scheduling/NextButton.tsx [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import React from "react";
+import { Button } from "@chakra-ui/react";
+import { NextButtonProps } from "./types";
+
+export default function NextButton({isBeingEdited, go, canSubmit, handleNext}: NextButtonProps) {
+  return (
+    <div style={{display: "flex", justifyContent: "flex-end"}}>
```
Comment (hanlinc27) [inline frontend/src/components/pages/Scheduling/NextButton.tsx:7]: nit: to adhere to ChakraUI
```suggestion
    <Flex justify="flex-end">
```

File: frontend/src/components/pages/Scheduling/SelectDateTime.tsx [hunk 8]
```diff
@@ -158,12 +174,15 @@ const SelectDateTime = ({
         ...formErrors,
         dayPart: "",
       });
+      checkSubmit(true);
     } else if (name === "frequency") {
       const val = e.toString();
-      if (val === "One time") {
+      if (val === "One time donation") {
```
Comment (hanlinc27) [inline frontend/src/components/pages/Scheduling/SelectDateTime.tsx:180]:
```suggestion
      if (val === Frequency.ONE_TIME) {
```

File: frontend/src/components/pages/Scheduling/SelectDateTime.tsx [hunk 9]
```diff
@@ -199,24 +219,41 @@ const SelectDateTime = ({
...
       ...formErrors,
       date: "",
     });
+
+    checkSubmit(false);
+
+    // update frequency labels
+    const newFrequencyLabels = [...frequencies];
+    newFrequencyLabels.forEach((freq, i) => {
+      if (freq === "Weekly") {
```
Comment (hanlinc27) [inline frontend/src/components/pages/Scheduling/SelectDateTime.tsx:243]:
```suggestion
      if (freq === Frequency.WEEKLY) {
```

File: frontend/src/components/pages/Scheduling/SelectDateTime.tsx [hunk 10]
```diff
@@ -199,24 +219,41 @@ const SelectDateTime = ({
...
+        newFrequencyLabels[i] = `Weekly on ${selectedDateObj.toLocaleString('en-us', {  weekday: 'long' })}s`;
+      } else if (freq === "Monthly") {
```
Comment (hanlinc27) [inline frontend/src/components/pages/Scheduling/SelectDateTime.tsx:245]:
```suggestion
      } else if (freq === Frequency.MONTHLY) {
```

File: frontend/src/components/pages/Scheduling/VolunteerInformation.tsx [hunk 11]
```diff
@@ -126,31 +131,34 @@ const VolunteerInformation = ({
...
-        <Text textStyle="mobileHeader4">Proposed dropoff time</Text>
-        <Text textStyle="mobileBody">{new Date(startTime).toDateString()}</Text>
+        <Text textStyle="mobileHeader4">Proposed Drop-off Time</Text>
+        <Text textStyle="mobileBody">{new Date(startTime).toLocaleDateString([], {weekday: 'long', month: 'long', day: 'numeric' })}</Text>
         <Text textStyle="mobileBody">
-          {new Date(startTime).toLocaleTimeString()}-
-          {new Date(endTime).toLocaleTimeString()}
+          {new Date(startTime).toLocaleTimeString([], {hour: '2-digit', minute:'2-digit'}).replace(/^(?:00:)?0?/, '')}-
+          {new Date(endTime).toLocaleTimeString([], {hour: '2-digit', minute:'2-digit'}).replace(/^(?:00:)?0?/, '')}
         </Text>
```
Comment (hanlinc27) [inline frontend/src/components/pages/Scheduling/VolunteerInformation.tsx:156]: in the future, could consider using the `date-fns` library we have for formatting
https://date-fns.org/
i.e.  https://github.com/uwblueprint/community-fridge-kw/blob/a5b84214a2532c7d52aa7bd99ca59ce37705ee4f/frontend/src/components/pages/Dashboard/components/DropoffCard.tsx#L23

**Review by hanlinc27 — CHANGES_REQUESTED** [review]
this is a really good first pr!!! @carissa-tang 
first pass comments, will test it out locally tomorrow or friday! :)

File: frontend/src/components/pages/Scheduling/DonationInformation.tsx (hunk 5 above)
Comment (xsharonhe) [reply frontend/src/components/pages/Scheduling/DonationInformation.tsx:65]: woah didn't know about this syntax :0 super cool

**Comment by carissa-tang** [comment]
Note: There is a bug rn with the "next" button where if an option was previously selected and repopulated, the button stays disabled (the code for changing the button currently only evaluates when there is an `onChange` event) - I haven't thought of the best solution for this yet... any thoughts are appreciated!

https://user-images.githubusercontent.com/69697180/150278983-494ebe11-40ca-4ccc-b70b-69ff4df534e1.mov

File: frontend/src/components/pages/Scheduling/SelectDateTime.tsx [hunk 12]
```diff
@@ -91,24 +95,29 @@ const SelectDateTime = ({
...
+    frequency === frequencies[0] || frequency === "",
   );
+  const [recurringEndDate, setRecurringEndDate] = useState<Date>(new Date(startTime));
 
   React.useEffect(() => {
     // Need to set frequency here since "One time"
     // is default value in desktop dropdown
     if (isDesktop && isOneTimeDonation) {
-      setForm({ target: { name: "frequency", value: "One time" } });
+      setForm({ target: { name: "frequency", value: frequencies[0] } });
```
Comment (xsharonhe) [inline frontend/src/components/pages/Scheduling/SelectDateTime.tsx:118]: this is probably more of a design note but i found it a little weird that all the other options were not automatically filled out but this one is - curious if there was a reason for this?

File: frontend/src/components/pages/Scheduling/SelectDateTime.tsx [hunk 13]
```diff
@@ -420,54 +447,28 @@ const SelectDateTime = ({
...
-          <Input
-            isDisabled={isBeingEdited}
-            value={recurringEndDate}
-            onChange={(e) => handleChangeRecurringDate(e.target.value)}
-            placeholder="MM/DD/YYYY"
-            maxWidth="350px"
-            size="lg"
-          />
+          <FormLabel fontWeight="400">Date</FormLabel>
+          <SimpleGrid columns={2} columnGap={16} rowGap={6} w="full">
```
Comment (xsharonhe) [inline frontend/src/components/pages/Scheduling/SelectDateTime.tsx:451]: is there any way to block out dates earlier than today? just out of curiousity for some potential errors to leak in

and we can't choose the next month - not sure if this is the wanted behaviour?

File: frontend/src/components/pages/Scheduling/NextButton.tsx [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+          onClick={handleNext}
+          variant="navigation"
+          width={{ md: "300px" }}
+        >
+          Next
+        </Button>
+      )}
+    </Flex>
+  )
+}
```
Comment (xsharonhe) [inline frontend/src/components/pages/Scheduling/NextButton.tsx:28]: nitpick: newline at end of file (you can see it on github too which files don't)

File: frontend/src/components/common/RadioImageSelectGroup.tsx [hunk 15]
```diff
@@ -111,7 +124,18 @@ const RadioImageSelectGroup = (props: RadioImageSelectGroupProps) => {
           <Text textStyle="mobileSmall">{v.description}</Text>
         </VStack>
       ) : (
-        <HStack>
+        <HStack style={{alignItems: "flex-start"}}>
```
Comment (xsharonhe) [inline frontend/src/components/common/RadioImageSelectGroup.tsx:127]: does style needed to be added here? can we just have the ```alignItems``` property

File: frontend/src/components/pages/Scheduling/SelectDateTime.tsx [hunk 16]
```diff
@@ -145,6 +154,13 @@ const SelectDateTime = ({
     setIcons(getIconsPerTimeSlot(selectedDayPart, selectedDate));
   };
 
+  const checkSubmit = (hasRecurringDonationEndDate: boolean) => {
+    const reccurring = hasRecurringDonationEndDate ? frequency && recurringDonationEndDate : true;
```
Comment (xsharonhe) [inline frontend/src/components/pages/Scheduling/SelectDateTime.tsx:158]: i think this is where your bug is occurring? maybe? i console.log(recurring) and it is giving me an undefined which will then affect your valuesFilled

File: frontend/src/components/pages/Scheduling/SelectDateTime.tsx (hunk 12 above)
Comment (xsharonhe) [reply frontend/src/components/pages/Scheduling/SelectDateTime.tsx:118]: i think this is part of the reason why the bug is introduced

File: frontend/src/components/pages/Scheduling/types.ts [hunk 17]
```diff
@@ -74,4 +90,4 @@ export const timeRanges = {
   night: ["9:00 PM - 10:00 PM", "10:00 PM - 11:00 PM", "11:00 PM - 12:00 AM"],
 };
 
-export const frequencies = ["One time", "Daily", "Weekly", "Monthly"];
+export const frequencies = ["One time donation", "Daily", "Weekly", "Monthly"];
```
Comment (xsharonhe) [inline frontend/src/components/pages/Scheduling/types.ts:93]: is there a reason why frequencies is an array here and not an enum? i noticed we already had an enum and i think it might be clearer since i can't really tell what ```frequencies[0]``` or ```frequencies[1]``` is clearly

**Review by xsharonhe — COMMENTED** [review]
Just some general thoughts 😄 Except for that one bug, everything is working as intended so that's fantastic! Really nice work 😸

File: frontend/src/components/pages/Scheduling/types.ts (hunk 17 above)
Comment (hanlinc27) [reply frontend/src/components/pages/Scheduling/types.ts:93]: not sure, @AmandaMisjuwar might be able to speak to it better but I think it might be getting mapped somewhere which is why it's an array?

File: frontend/src/components/pages/Scheduling/SelectDateTime.tsx (hunk 12 above)
Comment (carissa-tang) [reply frontend/src/components/pages/Scheduling/SelectDateTime.tsx:118]: i believe this was implemented because the "Frequency" input for desktop used to be a dropdown. But now they are radio buttons, so it should be fine if I remove it. Thanks for pointing that out!

File: frontend/src/components/pages/Scheduling/SelectDateTime.tsx (hunk 13 above)
Comment (hanlinc27) [reply frontend/src/components/pages/Scheduling/SelectDateTime.tsx:451]: yeah ideally the calendar is two weeks _starting_ today

**Review by hanlinc27 — APPROVED** [review]
